stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,553 @@
+	})
+	for _, key := range keys {
+		c.delete(key)
+	}
+}
+
+func (c *mergeJoinCache) Table(key flux.GroupKey) (flux.Table, error) {
+	entry, ok := c.data.Lookup(key)
+	if !ok {
+		return nil, errors.Newf(codes.NotFound, "no entry for group key %v in cache", key)
+	}
+	t := entry.(*cacheEntry)
+	if t.l == nil || t.r == nil {
+		return nil, errors.Newf(codes.NotFound, "no entry for group key %v in cache", key)
+	}
+	return c.joinTables(t.l, t.r)
+}
+
+func (c *mergeJoinCache) ForEach(f func(flux.GroupKey)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key)
+		}
+	})
+}
+
+func (c *mergeJoinCache) ForEachWithContext(f func(flux.GroupKey, execute.Trigger, execute.TableContext)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key, execute.NewTriggerFromSpec(c.spec), execute.TableContext{
+				Key: key,
+			})
+		}
+	})
+}
+
+func (c *mergeJoinCache) DiscardTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) ExpireTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) SetTriggerSpec(spec plan.TriggerSpec) {
+	c.spec = spec
+}
+
+func (c *mergeJoinCache) joinTables(a, b flux.Table) (flux.Table, error) {
+	// Compile row fn for the input rows
+	if err := c.fn.Prepare(a.Cols(), b.Cols()); err != nil {
+		return nil, err
+	}
+
+	key := a.Key()
+
+	// Create a table builder for the output of the join
+	builder := execute.NewColListTableBuilder(key, c.alloc)
+
+	ar := make([]flux.ColReader, 0)
+	br := make([]flux.ColReader, 0)
+
+	// Release column readers after joining
+	defer a.Done()
+	defer b.Done()
+
+	// Retain column readers for joining
+	if err := a.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		ar = append(ar, cr)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if err := b.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		br = append(br, cr)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	ca := execute.ColIdx(execute.DefaultTimeColLabel, a.Cols())
+	if ca < 0 {
+		return nil, errors.New(codes.NotFound, "no _time column found")
+	}
+
+	iterA := &colReaderIterator{
+		i:       -1,
+		timeCol: ca,
+		readers: ar,
+	}
+
+	cb := execute.ColIdx(execute.DefaultTimeColLabel, b.Cols())
+	if cb < 0 {
+		return nil, errors.New(codes.NotFound, "no _time column found")
+	}
+
+	iterB := &colReaderIterator{
+		i:       -1,
+		timeCol: cb,
+		readers: br,
+	}
+
+	rowA := make(map[string]values.Value, 8)
+	rowB := make(map[string]values.Value, 8)
+
+	_, ra, ta := iterA.next()
+	i, rb, tb := iterB.next()
+
+	var firstRow = true
+
+NEXT:
+	if ra == -1 || rb == -1 {
+		goto DONE
+	}
+
+	if ta < tb {
+		_, ra, ta = iterA.next()
+		goto NEXT
+	}
+	if ta > tb {
+		i, rb, tb = iterB.next()
+		goto NEXT
+	}
+
+	iterA.row(rowA)
+
+	for r, t := rb, tb; r != -1 && t == ta; _, r, t = iterB.next() {
+		iterB.row(rowB)
+
+		// Evaluate fn over both input rows
+		obj, err := c.fn.Eval(
```

Comment:
Can you move this to an `evalRow` function?

Reply:
Isn't this equivalent to `obj, err := evalRow(...)`?

Reply:
It wouldn't be very complex, but it would make it more explicit and allow you to pass `rowA` and `rowB` directly rather than the multiple lines needed for this invocation. I just found the function a bit difficult to read and I thought it would help to split it up into the various components.